update_channelsなどが動かない

```
{"ok"=>false, "error"=>"method_deprecated", "response_metadata"=>{"messages"=>["[ERROR] This method is retired and can no longer be used. Please use conversations.list or users.conversations instead. Learn more: https://api.slack.com/changelog/2020-01-deprecating-antecedents-to-the-conversations-api."]}}
```

conversations系のAPIに切り替える必要がある。

--- a/Gemfile
+++ b/Gemfile
@@ -1,6 +1,5 @@
 source 'https://rubygems.org'
 
-gem 'slack-api'
 gem 'sinatra'
 gem 'puma'
 gem 'rubyzip'
@@ -11,3 +10,6 @@ gem 'nokogiri'
 group :development do
   gem 'thin'
 end
+
+gem "slack-ruby-client"
+gem "async-websocket", '~> 0.8.0' # dependency of slack-ruby-client RTM

--- a/bin/init.rb
+++ b/bin/init.rb
@@ -12,10 +12,6 @@ config = {
     'secret_access_key' => nil,
   },
   'default_channel' => 'general',
-  'logger' => {
-    'enable_private_channel' => true,
-    'enable_direct_message' => true
-  },
   'database' => {
     'uri' => 'mongo:27017',
     'database' => 'slack_logger'

--- a/lib/db.rb
+++ b/lib/db.rb
@@ -25,15 +25,10 @@ def replace_channels(channels)
   end
 end
 
+# Ims
+#   for backward compatibility. should be removed.
 Ims = db['ims']
 Ims.indexes.create_one({ :id => 1 }, :unique => true)
-def replace_ims(ims)
-  unless ims.nil?
-    ids = ims.map{ |im| im['id'] }
-    Ims.find(id: { '$in' => ids }).delete_many
-    Ims.insert_many(ims)
-  end
-end
 
 Emojis = db['emojis']
 Emojis.indexes.create_one({ :name => 1 }, :unique => true)

--- a/lib/slack.rb
+++ b/lib/slack.rb
@@ -1,8 +1,13 @@
 require 'yaml'
-require 'slack'
+require 'slack-ruby-client'
 
 config = YAML.load_file('./config.yml')
 
 Slack.configure do |c|
   c.token = config['slack']['token']
+end
+
+Slack::RealTime::Client.configure do |c|
+  c.start_method = :rtm_connect
+  c.store_class = Slack::RealTime::Stores::Starter
 end

--- a/lib/slack_logger.rb
+++ b/lib/slack_logger.rb
@@ -3,10 +3,14 @@ require './lib/slack'
 require './lib/db'
 
 config = YAML.load_file('./config.yml')
-ENABLE_PRIVATE_CHANNEL = config['logger']['enable_private_channel']
-ENABLE_DIRECT_MESSAGE = config['logger']['enable_direct_message']
 
 class SlackLogger
+  attr_reader :client
+
+  def initialize
+    @client = Slack::Web::Client.new
+  end
+
   def is_private_channel(channel_name)
     channel_name[0] == 'G'
   end
@@ -16,37 +20,27 @@ class SlackLogger
   end
 
   def update_users
-    users = Slack.users_list['members']
+    users = client.users_list['members']
     replace_users(users)
   end
 
   def update_channels
-    channels = Slack.channels_list['channels']
+    channels = client.conversations_list({type: 'public_channel'})['channels']
     replace_channels(channels)
   end
 
-  def update_groups
-    groups = Slack.groups_list['groups']
-    replace_channels(groups)
-  end
-
-  def update_ims
-    ims = Slack.im_list['ims']
-    replace_ims(ims)
-  end
-
   def update_emojis
-    emojis = Slack.emoji_list['emoji']
+    emojis = client.emoji_list['emoji'] rescue nil
     replace_emojis(emojis)
   end
 
   # log history messages
   def fetch_history(target, channel)
-    messages = Slack.send(
+    messages = client.send(
       target,
       channel: channel,
       count: 1000,
-    )['messages']
+    )['messages'] rescue nil
 
     unless messages.nil?
       messages.each do |m|
@@ -58,13 +52,13 @@ class SlackLogger
 
   # realtime events
   def log_realtime
-    realtime = Slack.realtime
+    realtime = Slack::RealTime::Client.new
 
     realtime.on :message do |m|
-      if !ENABLE_PRIVATE_CHANNEL and is_private_channel(m['channel'])
+      if is_private_channel(m['channel'])
         next
       end
-      if !ENABLE_DIRECT_MESSAGE and is_direct_message(m['channel'])
+      if is_direct_message(m['channel'])
         next
       end
 
@@ -97,32 +91,13 @@ class SlackLogger
       update_emojis
     end
 
-    if ENABLE_PRIVATE_CHANNEL
-      realtime.on :group_joined do |c|
-        puts "group has joined"
-        update_groups
-      end
-
-      realtime.on :group_rename do |c|
-        puts "group has renamed"
-        update_groups
-      end
-    end
-
-    if ENABLE_DIRECT_MESSAGE
-      realtime.on :im_created do |c|
-        puts "direct message has created"
-        update_ims
-      end
-    end
-
     # if connection closed, restart the realtime logger
     realtime.on :close do
       puts "websocket disconnected"
       log_realtime
     end
 
-    realtime.start
+    realtime.start!
   end
 
   def start
@@ -134,24 +109,13 @@ class SlackLogger
       update_emojis
       update_users
       update_channels
-      update_groups if ENABLE_PRIVATE_CHANNEL
-      update_ims if ENABLE_DIRECT_MESSAGE
 
       Channels.find.each do |c|
         puts "loading messages from #{c[:name]}"
         if c[:is_channel]
-          fetch_history(:channels_history, c[:id])
-        elsif c[:is_group] && ENABLE_PRIVATE_CHANNEL
-          fetch_history(:groups_history, c[:id])
+          fetch_history(:conversations_history, c[:id])
         end
         sleep(1)
-      end
-
-      if ENABLE_DIRECT_MESSAGE
-        Ims.find.each do |i|
-          fetch_history(:im_history, i[:id])
-          sleep(1)
-        end
       end
 
       # realtime event is joined and dont exit current thread

--- a/viewer/viewer.rb
+++ b/viewer/viewer.rb
@@ -245,7 +245,8 @@ end
 get '/team.json' do
   content_type :json
   # TODO: cache in redis or mongodb or in memory?
-  Slack.team_info['team'].to_json
+  client = Slack::Web::Client.new
+  client.team_info['team'].to_json
 end
 
 post '/import_backup' do

